doc: (#68)
mongodb-labs/atlas-cli-plugin-terraform: docs/command_adv2v2.md

# Convert mongodbatlas_advanced_cluster to Provider 2.0.0 schema

>  advancedClusterToV2 (adv2v2) command helps you migrate previous `mongodbatlas_advanced_cluster` configurations to the new Provider 2.0.0 schema.

Reviewer: It could be helpful to clarify this.  ++"This revised file migrates the terraform configurations and state to the latest version and does not modify the resources deployed in MongoDB Atlas"

# Convert mongodbatlas_advanced_cluster to Provider 2.0.0 schema

 advancedClusterToV2 (adv2v2) command helps you migrate previous `mongodbatlas_advanced_cluster` configurations to the new Provider 2.0.0 schema.

This revised file migrates the Terraform configurations and state to the latest version and does not modify the resources deployed in MongoDB Atlas.

MongoDB Atlas Provider 2.0.0 introduces a new, cleaner structure for `mongodbatlas_advanced_cluster` resources. The main changes include the use of nested attributes instead of blocks and deletion of deprecated attributes like `disk_size_gb` at root level or `num_shards`.

## Usage

To convert a Terraform configuration from the previous `mongodbatlas_advanced_cluster` schema to the Provider 2.0.0 schema, use the following command:

```bash
atlas terraform advancedClusterToV2 --file in.tf --output out.tf
```

You can also use shorter aliases:
```bash
atlas tf adv2v2 -f in.tf -o out.tf
```

### Command Options

- `--file` or `-f`: Input file path containing the `mongodbatlas_advanced_cluster` configuration
- `--output` or `-o`: Output file path for the converted Provider 2.0.0 configuration

> - `--replaceOutput` or `-r`: Overwrite the output file if it exists, or even use the same output file as the input file

Reviewer 2: ```suggestion
- `--replaceOutput` or `-r`: Overwrite the file at the output path if it already exists. You can pass the input file path to this flag to modify the format in-place.
```
If that's the intended behavior, then I think this is clearest.

- `--watch` or `-w`: Keep the plugin running and watching for changes in the input file

## Examples

You can find [here](https://github.com/mongodb-labs/atlas-cli-plugin-terraform/tree/main/internal/convert/testdata/adv2v2) examples of input files (suffix .in.tf) and the corresponding output files (suffix .out.tf).

## Dynamic Blocks

`dynamic` blocks are used to generate multiple nested blocks based on a set of values. 
Given the different ways of using dynamic blocks, we recommend reviewing the output and making sure it fits your needs.

### Dynamic blocks in tags and labels

You can use `dynamic` blocks for `tags` and `labels`. The plugin assumes that `for_each` has an expression which is evaluated to a `map` of strings.

> You can also combine the use of dynamic blocks in `tags` and `labels` with individual blocks in the same cluster definition, e.g.:

Author: thanks, changed here: https://github.com/mongodb-labs/atlas-cli-plugin-terraform/pull/68/commits/3f5d2b2d24a84561dabe1010ad41e0deb7079512

This is an example of how to use dynamic blocks in `replication_specs`:
```hcl
dynamic "replication_specs" {
  for_each = var.replication_specs
  content {
    num_shards = replication_specs.value.num_shards
    zone_name  = replication_specs.value.zone_name # only needed if you're using zones
    dynamic "region_configs" {
      for_each = replication_specs.value.region_configs
      priority      = region_configs.value.priority
      provider_name = region_configs.value.provider_name
      region_name   = region_configs.value.region_name
      electable_specs {
        instance_size = region_configs.value.instance_size
        node_count    = region_configs.value.electable_node_count
      }
      # read_only_specs, analytics_specs, auto_scaling and analytics_auto_scaling can also be defined
    }
  }
}
```

### Limitations

If you need to use the plugin for `dynamic` block use cases not yet supported, please send us [feedback](https://github.com/mongodb-labs/atlas-cli-plugin-terraform/issues).

#### Combination of blocks with dynamic and inline expressions

> Dynamic blocks and individual blocks for `region_configs` or `replication_specs` are not supported at the same time. Remove the individual `region_configs` or `replication_specs` blocks and use a local `list` variable with [concat](https://developer.hashicorp.com/terraform/language/functions/concat) to add the individual block information to the variable you're using in the `for_each` expression.

Reviewer 3: ```suggestion
Dynamic blocks and individual blocks for `region_configs` or `replication_specs` are not supported at the same time. To continue using this plugin, remove the individual `region_configs` or `replication_specs` blocks and use a local `list` variable with [concat](https://developer.hashicorp.com/terraform/language/functions/concat) to add the individual block information to the variable you're using in the `for_each` expression.
```

> Let's see an example with `region_configs`; it's the same idea for `replication_specs`. In the original configuration file, the `mongodb_cluster` resource is used inside a module that receives the `region_configs` elements in a `list` variable and we want to add an additional `region_configs` with a read-only node.

Reviewer 3: ```suggestion
##### Example

Let's see an example with `region_configs`; the same idea applies for `replication_specs`. In the original configuration file, the `mongodbatlas_cluster` resource is used inside a module that receives the `region_configs` elements in a `list` variable and we want to add an additional `region_configs` with a read-only node.
```

> Let's see an example with `region_configs`, it is the same idea for `replication_specs`. In the original configuration file, the `mongodb_cluster` resource is used inside a module that receives the `region_configs` elements in a `list` variable and we want to add an additional `region_configs` with a read-only node.

Reviewer 2: ```suggestion
Let's see an example with `region_configs`; it is the same idea for `replication_specs`. In the original configuration file, the `mongodb_cluster` resource is used inside a module that receives the `region_configs` elements in a `list` variable and we want to add an additional `region_configs` with a read-only node.
```

  region_configs_all = concat(
    var.replication_specs.region_configs,
    [
      {
        priority        = 0
        provider_name   = "AWS"
        region_name     = "US_EAST_1"
        instance_size   = var.instance_size
        electable_node_count = 0
        read_only_node_count = 1
      },
    ]
  )
}

resource "mongodbatlas_advanced_cluster" "this" {
  project_id                  = var.project_id
  name                        = var.cluster_name
  cluster_type                = var.cluster_type
  replication_specs {
    num_shards = var.replication_specs.num_shards
    dynamic "region_configs" {
      for_each = local.region_configs_all # changed to use the local variable
      priority      = region_configs.value.priority
      provider_name = region_configs.value.provider_name
      region_name   = region_configs.value.region_name      
      electable_specs {
        instance_size = region_configs.value.instance_size
        node_count    = region_configs.value.electable_node_count
      }
      read_only_specs {
        instance_size = region_configs.value.instance_size
        node_count    = region_configs.value.read_only_node_count
      }
    }
  }
}
```

> This modified configuration file has the same behavior as the original one, but it doesn't have individual blocks anymore, only the `dynamic` block, so it is supported by the plugin.

Reviewer 2: ```suggestion
This modified configuration file has the same behavior as the original one, but it doesn't have individual blocks anymore, only the `dynamic` block, so it's supported by the plugin.
```
